repositories/rmw_fastrtps.BUILD.bazel

```diff
@@ -0,0 +1,48 @@
+""" Builds rmw_fastrtps.
+"""
+
+load("@com_github_mvukov_rules_ros2//ros2:cc_defs.bzl",
+    "ros2_cpp_binary", "ros2_cpp_library")
+
+ros2_cpp_library(
+    name = "rmw_fastrtps_shared_cpp",
+    srcs = glob([
+        "rmw_fastrtps_shared_cpp/include/**/*.h",
+        "rmw_fastrtps_shared_cpp/include/**/*.hpp",
+        "rmw_fastrtps_shared_cpp/src/*.cpp",
+        "rmw_fastrtps_shared_cpp/src/**/*.hpp",
+    ]),
+    includes = ["rmw_fastrtps_shared_cpp/include"],
+    deps = [
+        "@fastrtps",
+        "@ros2_rcpputils//:rcpputils",
+        "@ros2_rcutils//:rcutils",
+        "@ros2_rmw//:rmw",
+        "@ros2_rmw//:rmw_cpp",
+        "@ros2_rmw_dds_common//:rmw_dds_common_lib",
+        "@ros2_rosidl//:rosidl_runtime_c",
+        "@ros2_rosidl//:rosidl_typesupport_introspection_c",
+        "@ros2_rosidl//:rosidl_typesupport_introspection_cpp",
+        "@ros2_tracing//:tracetools",
+    ],
+)
+
+
+ros2_cpp_binary(
```

Comment:
So, fastdds supports two modes per https://fast-dds.docs.eprosima.com/en/latest/fastdds/ros2/ros2.html#ros-2-using-fast-dds-middleware:
* `rmw_fastrtps_dynamic_cpp` that can use introspection-based type-support that we already have. This one should be easy to integrate. cyclonedds rmw uses introspection-based type-support.
* `rmw_fastrtps_cpp` that uses their custom type-support. This is involved as fastdds-specific IDL code generation needs to be implemented.

This is for rclc/rclcpp part, I think. I'm not 100% sure, should be checked, which type-support rclpy expects/can work with.

My take: as a first step, if it suits your needs, just integrate rmw_fastrtps_dynamic_cpp.

If we want to go with rmw_fastrtps_cpp then we will also need to use dds_vendor switch in IDL code-generation, to only generate what's necessary. This might get involved -- haven't studied what generated code fastdds needs.

Please let me know what you think.

Reply:
I will switch it to dynamic instead. I tried this before, but didn't have improving results. Let's see if I can make it work!